lib/AST/GenericSignature.cpp

```diff
@@ -734,16 +715,17 @@ Type GenericSignatureImpl::getDependentUpperBounds(Type type) const {
           abort();
         }
 
-        types.push_back(ParameterizedProtocolType::get(ctx, baseType, argTypes));
+        types.push_back(ParameterizedProtocolType::get(getASTContext(),
+                                                       baseType, argTypes));
         continue;
       }
     }
-
-    types.push_back(baseType);
+    wantDependentBound ? types.push_back(proto->getDeclaredInterfaceType())
```

Comment:
If wantDependentBound is false, you can still use baseType if it doesn't contain type parameters.

Reply:
Also if it is false you should just skip the construction of the parameterized protocol types entirely.